nucypher/blockchain/eth/token.py

```diff
@@ -0,0 +1,196 @@
+from _pydecimal import Decimal
+
+import eth_utils
+import maya
+from eth_utils import currency
+from nacl.hash import sha256
+from typing import Union, Tuple
+
+from nucypher.blockchain.eth.agents import NucypherTokenAgent
+from nucypher.blockchain.eth.constants import TOKEN_DECIMALS
+from nucypher.blockchain.eth.utils import datetime_at_period, datetime_to_period
+
+
+class NU:
+    """
+    An amount of NuCypher tokens that doesn't hurt your eyes.
+    Wraps the eth_utils currency conversion methods.
+
+    The easiest way to use NU, is to pass an int, float, or str, and denomination string:
+
+    Int:    nu = NU(100, 'NU')
+    Int:    nu_wei = NU(15000000000000000000000, 'NuNit')
+
+    Float:  nu = NU(15042.445, 'NU')
+    String: nu = NU('10002.302', 'NU')
+
+    ...or alternately...
+
+    Float: nu = NU.from_tokens(100.50)
+    Int: nu_wei = NU.from_nu_wei(15000000000000000000000)
+
+    Token quantity is stored internally as an int in the smallest denomination,
+    and all arithmetic operations use this value.
+    """
+
+    __symbol = 'NU'
+    __decimals = TOKEN_DECIMALS
+    __agent_class = NucypherTokenAgent
+
+    # conversions
+    __denominations = {'NuNit': 'wei',
+                       'NU': 'ether'}
+
+    class InvalidAmount(ValueError):
+        """Raised when an invalid input amount is provided"""
+
+    def __init__(self, value: Union[int, float, str], denomination: str):
+
+        # Calculate smallest denomination and store it
+        wrapped_denom = self.__denominations[denomination]
+
+        # Convert or Raise
+        try:
+            self.__value = currency.to_wei(number=value, unit=wrapped_denom)
+        except ValueError as e:
+            raise NU.InvalidAmount(f"{value} is an invalid amount of tokens: {str(e)}")
+
+    @classmethod
+    def from_nunits(cls, value: int):
+        return cls(value, denomination='NuNit')
+
+    @classmethod
+    def from_tokens(cls, value: Union[int, float, str]):
+        return cls(value, denomination='NU')
+
+    def to_tokens(self) -> Decimal:
+        """Returns an decimal value of NU"""
+        return currency.from_wei(self.__value, unit='ether')
+
+    def to_nunits(self) -> int:
+        """Returns an int value in NuNit"""
+        return int(self.__value)
+
+    def __eq__(self, other) -> bool:
+        return int(self) == int(other)
+
+    def __radd__(self, other) -> 'NU':
```

Comment:
I think, `__radd__` and `__rmul__` are not needed when you have `__add__` and `__mul__`?

Reply:
I am entertaining the idea of using NU with `sum`, which, I think uses `__radd__`?